### ISSUE TITLE
Regression prediction mode

### ISSUE THREAD

**Issue by CrazyPython** [tqdm/tqdm#206]
I had a situation where every iteration, the time of the iteration slowed down. I propose a keyword argument to regress the time of each iteration to have better prediction time. It could simply do a exponential and linear regression and see which one fits better.

Like: `tqdm(..., regressmode=True)`

If this is to be implemented, there should be a warning about the possible slowdown.

**Comment by lrq3000** [comment]
This case is difficult and rare. The problem in implementing regressive methods is that they need to store all the past states (or at least a window), and `tqdm` philosophy of being fast and efficient implies that we use memoryless methods (we only know one saved value, the average, and the current value).

We talked about implementing an optional predictive/regressive model, but it is a great endeavour, and quite innovative (I didn't see any paper or software implementing such a model for progress bars or time prediction).

Initially, we intended to try to use machine learning, but it may be a bit too much.

Now I'm wondering if maybe we could use a memoryless regressive method, but I never heard of such a class of algorithm, a google scholar search should clarify that.

I'm now thinking such a memoryless estimator is possible because, when we think about it, what `tqdm` currently does is to estimate the average/min/max bounds using EMA.

Instead, we could maybe try to compute the vector of time progress, so we could know not only the value (how long it will take for next iteration), but also the direction (how much it will impact next next iteration time: constant, increasing or decreasing?).

If someone has heard of any algorithm that could fit here, I'd be glad to hear about it!

**Comment by CrazyPython** [comment]
@lrq3000 `tqdm.regress` as a separate one, so it doesn't slow down main operations? My case only had 24 iterations but it would take above 30 minutes (the first one is <1 minute). You can add a warning that such function is only for small iteration numbers. 

Just use the kind of simple regress method - the kind you find on a TI-84 calculator. and TI-84 has like 2 kHZ speed.

**Comment by casperdcl** [comment]
now we're talking. decent math rather than all this recent python implementation detail. Second order ema? Sounds v. promising, gonna give it a go this weekend :)

**Comment by casperdcl** [comment]
also, ti-84(+)? hello fellow ib learner!

**Comment by lrq3000** [comment]
Lol yeah I remember you are quite into data science @casperdcl :) I have absolutely no idea how a second order ema could be implemented, but yeah that would pretty much be the idea if you can do it :)

Otherwise, I have been thinking a bit about it. If we plot the cumulative time (current total time) as y and the number of iterations as x, we get something like this (warning: mspaint ahead):

![logistic-regression-tqdm-draft](https://cloud.githubusercontent.com/assets/1118942/17421099/6a3f34a6-5aa7-11e6-968f-b90924832479.png)

So this is a very good candidate for a logistic regression model. In addition, we get a function with interesting properties: bijective (so each (x,y) couple is unique in both directions), plus the function has limited degrees of freedom: it can only increase, with extremes being constant or vertical. I don't know if we can include these constraints in the logistic regression (maybe as a form of regularization penalizing decreasing slopes?), but it's surely something we can use to make a more precise predictive model.

Also, we can use an online logistic regression, so we would still be memoryless, plus if we add a discount factor we can get something alike the `smoothing` parameter we have.

Now I did a bit of googling, but it seems that making logistic regression online is a bit more tricky than what one can guess, so I'm not sure we can implement it (since it's still a topic of research). More googling is necessary.

**Comment by lrq3000** [comment]
Arg, I didn't really explain why the cumulative time/iterations model is a good model for our purpose:

1. We already have both variables in stores, we can use them easily.
2. Our target is to predict the total time, and here it's quite direct: our model will learn y, which is the cumulative time, in other words the total time taken to reach iteration x. To predict the end total time, we then just have to predict y for x=last iteration. Easy and fast.

**Comment by lrq3000** [comment]
Please see and test #248 and tell me if it fits your needs.

**Comment by lrq3000** [comment]
If it fits the bill, I'll finish the PR.

**Comment by almson** [comment]
A second-order EMA will graph a parabola. In certain common cases, it will predict a negative iteration duration and an imaginary completion time. (Something to ponder?)

Unless tqdm remembers previous runs of your script, nothing short of General AI is going to do a much better job of predicting completion time than an average.

**Comment by lrq3000** [comment]
> Unless tqdm remembers previous runs of your script, nothing short of
General AI is going to do a much better job of predicting completion time
than an average.

I disagree. The average will always result in a linear progress rate (I'm
simplifying since we use EMA which can allow for some dynamic adjustment
and hence non linearity, but it's ultimately just several linear segments
stitched together).

The point of the regression mode is to try to catch non linear patterns and
to account for them to provide a better time estimate.

One concrete example is a task where each ith iteration can take longer
than the others, let's say every 3 iterations. With EMA, tqdm will always
underestimate the time that every 3rd iterations take, but on the 4th it
will overestimate due to the 3rd. Then on the 5th it will be readjusted
given the fast 4th, and then on the slow 6th iteration it will
underestimate again, etc.

With a regression (or more complex models, i considered making a more
general api to plug in sci-kit learn), this could be accounted and IIRC i
tested such kind of scenario.

Since it's computationally more intensive, i thought the regression mode
would be more useful for scientific applications eg for the higher level
loops, i don't expect such approach to be useful in production where speed
is more important than precision of remaining time calculation.

**Comment by almson** [comment]
Umm, a simple case of the 4th iteration taking longer is exactly what a (correctly implemented) average takes care of. Tqdm had bugs up to 4.55 but now it's fixed.

**Comment by lrq3000** [comment]
You're correct, this case was already covered in tqdm even since its conception by disabling EMA to be in SMA-only mode. However, let's imagine your program's iterations pattern can change between launches, eg network application, then the SMA-only mode cannot cope with that accurately and is the EMA was implemented to more dynamically readjust. I wasn't aware of the changes post-4.55 so maybe both cases are more elegantly handled without requiring user interaction to switch between modes depending on the expected application, but I doubt they are accurately handled, as these limitations are intrinsic to the models used.

/EDIT: in fact my example was off, to quote my self from the past: "normal tqdm is exponential moving average, so it cannot predict at all when iterations are taking longer and longer because it assumes constance."

Predicting the time a process will take to complete is a non-trivial task. The weighted combination of SMA+EMA is a simple model, that works in the vast majority of cases good enough for the needs, but is not necessarily accurate especially in some edge cases. There are lots of other models for time prediction, and this issue/[my PR](https://github.com/tqdm/tqdm/pull/248) was just a proof-of-concept to open the gates for more advanced, although niche, models, where accurate remaining time prediction may be more important than resource consumption.

I also think it's an interesting scientific experiment as it's fun to think that tqdm can be used as a measuring instrument to collect  data that we can then model :-)
